examples/puzzles/river/river.qnt

```diff
@@ -0,0 +1,112 @@
+// -*- mode: Bluespec; -*-
+/**
+ * This is an example of a river-crossing puzzle [1].
+ * The specification itself is quite simple, so it can serve as a good
+ * starting example.
+ *
+ * [1]: https://en.wikipedia.org/wiki/River_crossing_puzzle
+ *
+ * "The fox, goose and bag of beans puzzle, in which a farmer must transport a
+ * fox, goose and bag of beans from one side of a river to another using a boat
+ * which can only hold one item in addition to the farmer, subject to the
+ * constraints that the fox cannot be left alone with the goose, and the goose
+ * cannot be left alone with the beans. Equivalent puzzles have also been stated
+ * involving a fox, chicken, and bag of grain, or a wolf, goat, and cabbage, etc."
+ *
+ * Igor Konnov, Informal Systems, 2023.
+ */
+module river {
+  // who is on the origin bank, a subset of "wolf", "goat", "cabbage", "farmer"
+  var origin: Set[str]
+  // what is on the boat, in addition to the farmer
+  var boat: Set[str]
+  // who is on the destination bank
+  var destination: Set[str]
+
+  // the safety property that must be enforced in the puzzle
+  pure def isSafe(location: Set[str]): bool = and {
+    Set("goat", "cabbage").subseteq(location) implies location.contains("farmer"),
+    Set("goat", "wolf").subseteq(location) implies location.contains("farmer"),
+  }
+
+  // we start with everybody on the origin bank and empty boat
+  action init = all {
+    origin' = Set("wolf", "goat", "cabbage", "farmer"),
+    boat' = Set(),
+    destination' = Set(),
+  }
+
+  // when the farmer is on the origin bank, they can bring at most one item
+  action embarkOrigin = all {
+    origin.contains("farmer"),
+    nondet toEmbark =
+      powerset(origin)
+        .filter(s => size(s) <= 2 and "farmer".in(s))
+        .oneOf()
```

Comment:
For my own understanding, does the simulator construct the powerset explicitly in this case? (it's fine here, since `|origin| <= 4`)

Reply:
Yeah, especially, in the combination with `filter`, it is definitely constructing the powerset. Actually, I first wrote it symbolically, but that was too bad for the simulator.